### ISSUE TITLE
Invalid option in dhcp.conf, breaking dhcpd service on RHEL6

### ISSUE THREAD

**Issue by russauld** [xcat2/xcat-core#1994]
In commit ee1a1574c361ce85b30c04b118c12fe611607bf9 from Sept 2014, the following line was added to dhcp.pm:

```
push @netent,
          "    } else if option client-architecture = 00:0e { #OPAL-v3\n ";
        push @netent, "      option conf-file = \"http://$tftp/tftpboot/pxelinux.cfg/p/" . $net . "_" . $maskbits . "\";\n";
        push @netent,
```

The trouble is that "option conf-file" is invalid on RHEL6.7. The resultant dhcpd.conf file results in a dhcpd service that will not start.

```
dhcpd: /etc/dhcp/dhcpd.conf line 848: no option named conf-file in space dhcp
dhcpd:        option conf-file =
dhcpd:                ^
dhcpd: /etc/dhcp/dhcpd.conf line 849: expecting a parameter or declaration
dhcpd:     } else
dhcpd:      ^
dhcpd: Configuration file errors encountered -- exiting
```

It would appear that the service is parsing the entire file, despite the effort to wrap the code in an if check for client-architecture.
The offending line can be commented out manually and the file will otherwise function properly. Subsequent calls to "makedhcp" will not change the section back since it is already present for the given interface.

I'm surprised that no one else has had this problem given that this code is two years old.

**Comment by immarvin** [comment]
hi @penguhyang , would you please recreate this issue in our env? 
install xCAT on a fresh redhat6.7 code and run ``service dhcpd restart``

**Comment by penguhyang** [comment]
This issue can't be recreated.

**Comment by immarvin** [comment]
hi @russauld ，

would you please provide the /etc/dhcp/dhcpd.conf on you env? since "conf-file" is the user defined option name for dhcp option 209.  The definition is in the beginning of  /etc/dhcp/dhcpd.conf
````
[root@c910f02c09p07 /]# cat /etc/dhcp/dhcpd.conf
#xCAT generated dhcp configuration

option conf-file code 209 = text;
...
````
so the error ``no option named conf-file in space dhcp`` might be caused by the fact that this conf is not defined in /etc/dhcp/dhcpd.conf. please make sure this line exists in your /etc/dhcp/dhcpd.conf. 

 

we can not recreate this issue on redhat 6.7 with latest xCAT.
````
[root@c910f02c09p07 /]# cat /etc/*release
Red Hat Enterprise Linux Server release 6.7 (Santiago)
Red Hat Enterprise Linux Server release 6.7 (Santiago)
[root@c910f02c09p07 /]# uname -a
Linux c910f02c09p07 2.6.32-573.el6.ppc64 #1 SMP Wed Jul 1 18:21:11 EDT 2015 ppc64 ppc64 ppc64 GNU/Linux
[root@c910f02c09p07 /]# rpm -qa|grep dhcpd
[root@c910f02c09p07 /]# rpm -qa|grep dhcp
dhcp-common-4.1.1-49.P1.el6.ppc64
dhcp-4.1.1-49.P1.el6.ppc64
[root@c910f02c09p07 /]# which dhcpd
/usr/sbin/dhcpd
[root@c910f02c09p07 /]# rpm -qf /usr/sbin/dhcpd
dhcp-4.1.1-49.P1.el6.ppc64
[root@c910f02c09p07 /]# lsxcatd -v
Version 2.12.4 (git commit 60d9b5fd9ea706e1212a977dc5c766aee47100a4, built Wed Oct 19 06:16:19 EDT 2016)
[root@c910f02c09p07 /]#
````

**Comment by russauld** [comment]
I removed the existing "dhcpd.conf" file and regenerated a new one with "makedhcp -n".
That resolved the issue.
So the problem really is that changes to the Perl module combined with not completely regenerating the configuration file resulted in a incorrect stanza when a new interface was added to the system (virbr1 for virtualization).
The dhcpd.conf file that we had was generated with an older version of xCAT (v2.8.4 or earlier). Since the Perl code will not touch anything it doesn't need to updated, it only added the new section for the virbr1 interface and that had the extra "option" field in it.
Is there a recommendation to completely regenerate "dhcpd.conf" when a xCAT is updated?
Is that a best practice? Are there any other files that might be similarly affected?
Is this documented anywhere?

I'll close this issue. Sorry for the inconvenience.

**Comment by immarvin** [comment]
hi @russauld , ``makedhcp -n`` will always backup the original /etc/dhcp/dhcpd.conf to /etc/dhcp/dhcpd.conf.xcatbak , and regenerate the new dhcpd.conf.

**Comment by russauld** [comment]
I understand that part, but what is the expectation by the developers?
Is it best practice to run `makedhcp -n` every time I update xCAT?

**Comment by immarvin** [comment]
hi @russauld , you expect the "/etc/dhcpd/dhcpd.conf" be updated on each time xCAT be upgraded, I think this is a reasonable requirement, we can consider this. Please trace the update on this ticket. thanks

**Comment by whowutwut** [comment]
A pull request by @immarvin  needs to be merged before we can mark this pending, moving to 2.13 release as this fix is not in the 2.12.4 GA today.
